### PR TITLE
Respect `min` and `max` of inputs to create more precise repro scripts

### PR DESCRIPTION
**Pull request by crcrpar**
Related: #4529

When an nvfuser fusion definition takes signed int tensors as inputs, it's possible that those int tensors have a solid value range e.g. [0, some_large_number) for embedding, the `FusionDefinition.last_repro_script` seems to have lost such range information as it uses `FusionDefinition.fake_inputs` that is generated by casting all the input tensors, leading to a loss of actual values.

Inspired by lightning-thunder's [`ExampleInputMetaData`](https://github.com/Lightning-AI/lightning-thunder/blob/4741b440d0e9e187b765e7dd9b0a23b3c2b6d7bc/thunder/dynamo/utils.py#L89), I rather store the range info in a dataclass that can give us a string of `make_tensor` or equivalents to get tensors of the same value range.

For the same fusion definition as #4529, running the fusion definition as follows
```python
inputs = [
    torch.randint(0, 10, (16,), dtype=torch.int32, device='cuda:0'),
    torch.testing.make_tensor((151936, 2048), dtype=torch.bfloat16, device='cuda:0'),
    torch.testing.make_tensor((2048,), dtype=torch.bfloat16, device='cuda:0'),
    torch.testing.make_tensor((2560, 2048), dtype=torch.bfloat16, device='cuda:0'),
    torch.testing.make_tensor((2560,), dtype=torch.bfloat16, device='cuda:0'),
]
fd.execute(inputs, save_repro_inputs=True, print_repro=True)
```
both `fd.last_repro_script()` and the `fd.execute` return
```python
inputs = [
    torch.testing.make_tensor((16,), dtype=torch.int32, device="cuda:0", low=0, high=7,),
    torch.testing.make_tensor((151936, 2048), dtype=torch.bfloat16, device="cuda:0", low=-9.0, high=8.9375,),
    torch.testing.make_tensor((2048,), dtype=torch.bfloat16, device="cuda:0", low=-9.0, high=8.9375,),
    torch.testing.make_tensor((2560, 2048), dtype=torch.bfloat16, device="cuda:0", low=-9.0, high=8.9375,),
    torch.testing.make_tensor((2560,), dtype=torch.bfloat16, device="cuda:0", low=-9.0, high=8.9375,),
]
```

### CODE REVIEW

File: python/nvfuser/__init__.py [hunk 1]
```diff
@@ -55,6 +56,51 @@ def disable_automatic_serialization():
...
+    low: Number
+    high: Number
+    device: str
+    dtype: torch.dtype
+    size: tuple[int, ...]
+    strides: tuple[int, ...]
+    is_contiguous: bool
+
+    def __init__(self, tensor: torch.Tensor) -> None:
+        if type(tensor) is not torch.Tensor:
```
[the remark this answers is not in the document]
Comment (crcrpar) [reply python/nvfuser/__init__.py:71]: Here I want to make sure tensor is compatible with `torch.aminmax`. So what I really want to filter out is tensor subclasses that we cannot reconstruct with `torch.testing.make_tensor`, `torch.rand`, or `torch.randint`. So in short `isinstance` wouldn't be sufficing

File: python/nvfuser/__init__.py [hunk 2]
```diff
@@ -311,10 +356,7 @@ def execute(
                     self._finalize_schedule(inputs)
 
         if save_repro_inputs:
-            from torch._subclasses.fake_tensor import FakeTensorMode
-
-            fake_mode = FakeTensorMode()
-            self.fake_inputs = [fake_mode.from_tensor(inp) for inp in inputs]
+            self.last_input_reproducers = [InputReproducer(tensor) for tensor in inputs]
```
[the remark this answers is not in the document]
Comment (crcrpar) [reply python/nvfuser/__init__.py]: I chose not to remove `fake_inputs` and to add `last_input_reproducers` with storage_offset and requires_grad

**Comment by naoyam** [comment]
@rdspring1 Can you please review this PR?

**Comment by rdspring1** [comment]
Is this ready-for-review?

**Comment by crcrpar** [comment]
Yes

File: python/nvfuser/__init__.py [hunk 3]
```diff
@@ -519,28 +582,11 @@ def repro_script_for(self, inputs: list | None = None) -> str:
...
-                            )
-                        else:
-                            upper_bound = 2 if i.dtype == torch.bool else 10
-                            msg += (
-                                f"    torch.randint(0, {upper_bound}, ({sz},), dtype={i.dtype}, device='{i.device}')"
-                                f".as_strided({tuple(i.size())}, {tuple(i.stride())}),\n"
-                            )
+                    i = InputDescriptor(i)
+                if isinstance(i, InputDescriptor):
+                    msg += f"    {i.to_make_tensor_str()},\n"
```
Comment (rdspring1) [inline python/nvfuser/__init__.py]: Nice!

File: python/nvfuser/__init__.py [hunk 4]
```diff
@@ -55,6 +56,63 @@ def disable_automatic_serialization():
...
+        return torch.testing.make_tensor(
+            self.size,
+            dtype=self.dtype,
+            device=self.device,
+            low=self.low,
+            high=self.high,
+            requires_grad=self.requires_grad,
+        ).as_strided(self.size, self.strides, self.storage_offset)
+
+    def to_make_tensor_str(self) -> str:
```
Comment (rdspring1) [inline python/nvfuser/__init__.py]: Why not use the `__str__` method?
```suggestion
    def __str__(self) -> str:
```

File: python/nvfuser/__init__.py [hunk 5]
```diff
@@ -55,6 +56,63 @@ def disable_automatic_serialization():
...
+            return 0, 2
+
+        t = tensor
+        if t.dtype.is_floating_point and t.dtype.itemsize == 1:
+            t = t.float()
+
+        min_max = torch.aminmax(t)
+        return min_max[0].cpu().item(), min_max[1].cpu().item()
+
+    def make_repro_tensor(self) -> torch.Tensor:
```
Comment (rdspring1) [inline python/nvfuser/__init__.py]: How is this used or tested?

Comment (rdspring1) [reply python/nvfuser/__init__.py]: Could this function become `__call__`? Then, `InputDescriptor` becomes an `InputTensorFactory`.

**Review by rdspring1 — APPROVED** [review]
LGTM. Left some random ideas.

File: python/nvfuser/__init__.py (hunk 5 above)
Comment (rdspring1) [reply python/nvfuser/__init__.py]:
```python
factory = InputTensorFactory(tensor)
fake_input = factory()
run_something(fake_input)
```

Comment (crcrpar) [reply python/nvfuser/__init__.py]: As for the use, yes, I think this can be YAGNI. I'll remove it for now. Once Thunder starts using this class, then there can be a need.

**Comment by crcrpar** [comment]
@rdspring1 could you take another look at this?

File: python/nvfuser/__init__.py [hunk 6]
```diff
@@ -310,11 +358,13 @@ def execute(
                     self.schedule()
                     self._finalize_schedule(inputs)
 
+        # TODO(crcrpar): Think about deprecating `self.fake_inputs` in favor of `InputDescriptor`.
```
Comment (rdspring1) [inline python/nvfuser/__init__.py:364]: What are the tradeoffs between using `self.fake_inputs` and `self.input_descriptors`?

File: python/nvfuser/__init__.py [hunk 7]
```diff
@@ -522,28 +575,11 @@ def repro_script_for(self, inputs: list | None = None) -> str:
...
-                                f"    torch.randn({sz}, dtype={i.dtype}, device='{i.device}')"
-                                f".as_strided({tuple(i.size())}, {tuple(i.stride())}),\n"
-                            )
-                        else:
-                            upper_bound = 2 if i.dtype == torch.bool else 10
-                            msg += (
-                                f"    torch.randint(0, {upper_bound}, ({sz},), dtype={i.dtype}, device='{i.device}')"
-                                f".as_strided({tuple(i.size())}, {tuple(i.stride())}),\n"
-                            )
+                    i = InputTensorFactory(i)
```
Comment (rdspring1) [inline python/nvfuser/__init__.py:583]: So, this part converts `torch.Tensor` to `InputTensorFactory`, then uses `InputTensorFactory` `__str__` function for a complete tensor definition.

**Review by rdspring1 — APPROVED** [review]
LGTM

File: python/nvfuser/__init__.py (hunk 6 above)
Comment (crcrpar) [reply python/nvfuser/__init__.py:364]: `fake_inputs` could lose potentially critical information of value ranges of the corresponding tensor. e.g. int64_t index tensors should not have negative values but fake inputs could lose it.
